### PR TITLE
ENH: Add dtype argument to read_sql_query (GH10285)

### LINKED ISSUE #10285: Add dtype keyword to read_sql_query to control per column dtypes.
When reading from SQL queries - especially when reading chunk-wise - issues with type conversion can occur.

It would be ideal if one could specify the types of the columns, so that type conversions go right.

A very common case is reading pure float results chunk-wise from a large table. The nuisance comes in when and chunk contains NULL values only in a column. In such cases these NULL values are stored as `None` rather than  `float("nan")`. I see that this is an issue with inconsistent NULL types in pandas and missing type information in read_sql_query.
Another case could be when text columns happen to contain numbers only in a chunk.

Specifying types for a query manually could resolve these issues.

### PR DESCRIPTION
**Pull request by avinashpancham**
- [x] closes #10285 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by avinashpancham** [comment]
@jorisvandenbossche tagging you since you were involved in the original issue. Could you have a look at this imp and lmk whether this is what you had in mind? If the imp is fine then I will add the tests as well.

**Comment by avinashpancham** [comment]
@jorisvandenbossche could you (or maybe someone else) have a look at this PR? Thanks

**Comment by avinashpancham** [comment]
@jreback could you (or maybe someone else) have a look at this PR? Thanks

**Review by jreback — CHANGES_REQUESTED** [review]
pls *always* add tests. w/o them its impossible to tell if this does *anything*

**Comment by avinashpancham** [comment]
Updated PR. Only Linux py38_np_dev fails, but due to other reasons.

**Comment by avinashpancham** [comment]
Merged master, everything succeeds.

I also saw a similar enhancement request for dtypes for `read_sql`. Once we agree on this imp we can easily extend it to that function as well

**Comment by avinashpancham** [comment]
@jreback could you have another look? Will fix the merge conflict in the docs later today

File: pandas/io/sql.py [hunk 1]
```diff
@@ -295,6 +301,7 @@ def read_sql_query(
     params=None,
     parse_dates=None,
     chunksize: None = None,
+    dtype: Optional[Union[Dtype, Dict[str, Dtype]]] = None,
```
Comment (jreback) [inline pandas/io/sql.py:304]: this type actually is pretty usefule, can you define in _typing, call it DtypeOrDictDtype / DtypeTable and add a comment about it. cc @simonjayhawkins @WillAyd @jorisvandenbossche for the name here.

File: pandas/io/sql.py [hunk 2]
```diff
@@ -1225,7 +1238,13 @@ def read_table(
...
     @staticmethod
     def _query_iterator(
-        result, chunksize, columns, index_col=None, coerce_float=True, parse_dates=None
+        result,
+        chunksize,
+        columns,
+        index_col=None,
+        coerce_float=True,
+        parse_dates=None,
+        dtype=None,
```
Comment (jreback) [inline pandas/io/sql.py:1247]: can you type anywhere you are adding type

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -307,6 +307,7 @@ Other enhancements
 - Improve numerical stability for :meth:`.Rolling.skew`, :meth:`.Rolling.kurt`, :meth:`Expanding.skew` and :meth:`Expanding.kurt` through implementation of Kahan summation (:issue:`6929`)
 - Improved error reporting for subsetting columns of a :class:`.DataFrameGroupBy` with ``axis=1`` (:issue:`37725`)
 - Implement method ``cross`` for :meth:`DataFrame.merge` and :meth:`DataFrame.join` (:issue:`5401`)
+- :func:`pandas.read_sql_query` now accepts a ``dtype`` argument to cast the columnar data from the SQL database based on user input (:issue:`10285`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:310]: move to 1.3

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/sql.py (hunk 1 above)
Comment (WillAyd) [reply pandas/io/sql.py:304]:
```suggestion
    dtype: Optional[Union[Dtype, Dict[Label, Dtype]]] = None,
```

Maybe `DtypeArg`?

File: pandas/_typing.py [hunk 4]
```diff
@@ -99,6 +92,14 @@
...
 
+# dtypes
+
+Dtype = Union[
+    "ExtensionDtype", str, np.dtype, Type[Union[str, float, int, complex, bool, object]]
+]
+DtypeArg = Optional[Union[Dtype, Dict[Label, Dtype]]]
+DtypeObj = Union[np.dtype, "ExtensionDtype"]
+
 # For functions like rename that convert one label to another
```
Comment (avinashpancham) [inline pandas/_typing.py:103]: Moved  the dtype block since `Label` was defined later in the script

File: pandas/io/sql.py (hunk 1 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done, added `DtypeArg` to _typing

File: pandas/io/sql.py (hunk 2 above)
Comment (avinashpancham) [reply pandas/io/sql.py]: Done

File: doc/source/whatsnew/v1.2.0.rst (hunk 3 above)
Comment (avinashpancham) [reply doc/source/whatsnew/v1.2.0.rst]: Done

**Comment by avinashpancham** [comment]
In #13049 it is proposed to add the `dtype` argument to `read_sql`. This can be done in similar manner as in this PR.

File: pandas/_typing.py [hunk 5]
```diff
@@ -100,6 +93,14 @@
 JSONSerializable = Optional[Union[PythonScalar, List, Dict]]
 Axes = Collection
 
+# dtypes
+
+Dtype = Union[
+    "ExtensionDtype", str, np.dtype, Type[Union[str, float, int, complex, bool, object]]
+]
+DtypeArg = Optional[Union[Dtype, Dict[Label, Dtype]]]
```
Comment (jreback) [inline pandas/_typing.py:101]: don't use Optional in the spec itself

File: pandas/io/sql.py [hunk 6]
```diff
@@ -132,10 +133,14 @@ def _wrap_result(
     index_col=None,
     coerce_float: bool = True,
     parse_dates=None,
+    dtype: DtypeArg = None,
```
Comment (jreback) [inline pandas/io/sql.py:136]: Optional[DtypeArg] for all of these

File: pandas/_typing.py [hunk 7]
```diff
@@ -100,6 +93,14 @@
 JSONSerializable = Optional[Union[PythonScalar, List, Dict]]
 Axes = Collection
 
+# dtypes
+
+Dtype = Union[
```
Comment (jreback) [inline pandas/_typing.py:98]: can you add a comment on what DtypeArg is / supposed to be used

File: pandas/io/sql.py [hunk 8]
```diff
@@ -1361,6 +1376,9 @@ def read_query(
         chunksize : int, default None
             If specified, return an iterator where `chunksize` is the number
             of rows to include in each chunk.
+        dtype : Type name or dict of columns
+            Data type for data or columns. E.g. np.float64 or
+            {‘a’: np.float64, ‘b’: np.int32, ‘c’: ‘Int64’}
```
Comment (jreback) [inline pandas/io/sql.py:1381]: versionadded 1.3

File: pandas/tests/io/test_sql.py [hunk 9]
```diff
@@ -937,6 +937,25 @@ def test_multiindex_roundtrip(self):
...
+            (float, [float, float]),
+            ({"SepalLength": int, "SepalWidth": float}, [int, float]),
+        ],
+    )
+    def test_dtype_argument(self, dtype, expected):
+        # GH10285 Add dtype argument to read_sql_query
+        result = sql.read_sql_query(
+            "SELECT SepalLength, SepalWidth FROM iris", self.conn, dtype=dtype
+        )
+        assert result.dtypes.to_dict() == {
```
Comment (jreback) [inline pandas/tests/io/test_sql.py:954]: can you constructed an expected frame and use tm.assert_frame_equal

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm ping on greenish

**Comment by jreback** [comment]
as a followup :-> can type any `dtype=` args in the codebase

**Comment by avinashpancham** [comment]
@jreback CI is greenish,Travis failed due to other reasons

Will work on dtype follow up tomorrow.

File: pandas/io/sql.py [hunk 10]
```diff
@@ -371,6 +379,9 @@ def read_sql_query(
     chunksize : int, default None
         If specified, return an iterator where `chunksize` is the number of
         rows to include in each chunk.
+    dtype : Type name or dict of columns
+        Data type for data or columns. E.g. np.float64 or
+        {‘a’: np.float64, ‘b’: np.int32, ‘c’: ‘Int64’}
```
Comment (jreback) [inline pandas/io/sql.py:384]: need a versionadded 1.3 here. ok to add in next PR

**Comment by jreback** [comment]
thanks @avinashpancham 

small comment for followon

File: pandas/io/sql.py (hunk 10 above)
Comment (avinashpancham) [reply pandas/io/sql.py:384]: Sorry, I see i didn't commit that change. But will indeed add it to the follow on

**Pull request merged**